python/ray/experimental/channel.py

```diff
@@ -187,3 +188,155 @@ def close(self) -> None:
+
+    def close(self) -> None:
+        self._closed = True
+        for channel in self._input_channels:
+            channel.close()
+
+
+class SynchronousInputReader(InputReader):
+    def __init__(self, input_channels: List[Channel]):
+        super().__init__(input_channels)
+
+    def start(self):
+        pass
+
+    def _begin_read_list(self) -> Any:
+        return [c.begin_read() for c in self._input_channels]
+
+    def end_read(self) -> Any:
+        for c in self._input_channels:
+            c.end_read()
+
+
+class AwaitableBackgroundOutputReader(InputReader):
+    def __init__(self, input_channels: List[Channel], _fut_queue: asyncio.Queue):
+        super().__init__(input_channels)
+        self._fut_queue = _fut_queue
+        self._background_task = None
+
+    def start(self):
+        self._background_task = asyncio.ensure_future(self.run())
+
+    def _run(self):
+        vals = [c.begin_read() for c in self._input_channels]
+        if self._has_single_output:
+            vals = vals[0]
+        return vals
+
+    async def run(self):
+        loop = asyncio.get_running_loop()
+        while not self._closed:
+            res, fut = await asyncio.gather(
+                loop.run_in_executor(None, self._run),
+                self._fut_queue.get(),
+                return_exceptions=True,
+            )
+
+            fut.set_result(res)
+
+    def end_read(self) -> Any:
+        for c in self._input_channels:
+            c.end_read()
+
+    def close(self):
+        self._background_task.cancel()
+        super().close()
+
+
+class OutputWriter:
+    def __init__(self, output_channel: Channel):
+        self._output_channel = output_channel
+        self._closed = False
+        self._num_writes = 0
+
+    def get_num_writes(self) -> int:
+        return self._num_writes
+
+    def start(self):
+        raise NotImplementedError()
+
+    def write(self, val: Any) -> None:
+        raise NotImplementedError()
+
+    def close(self) -> None:
+        self._closed = True
+        self._output_channel.close()
+
+
+class SynchronousOutputWriter(OutputWriter):
+    def start(self):
+        pass
+
+    def write(self, val: Any) -> None:
+        self._output_channel.write(val)
+        self._num_writes += 1
+
+
+class AwaitableBackgroundOutputWriter(OutputWriter):
+    def __init__(self, output_channel: Channel, max_queue_size: Optional[int] = None):
+        super().__init__(output_channel)
+        if max_queue_size is None:
+            max_queue_size = 0
+        self._queue = asyncio.Queue(max_queue_size)
+        self._background_task = None
+
+    def start(self):
+        self._background_task = asyncio.ensure_future(self.run())
+
+    def _run(self, res):
+        self._output_channel.write(res)
+
+    async def run(self):
+        loop = asyncio.get_event_loop()
+        while True:
+            res = await self._queue.get()
+            await loop.run_in_executor(None, self._run, res)
```

Comment:
I think None will create the threadpool which uses total_num_cpus / 2 threads by default. Should we create a threadpool with a single thread here instead?